Automate KubeCon event listings

Caesarsage opened the pull request:
### Description
Replaces hardcoded KubeCon event links with an automated system that fetches and displays events from the Linux Foundation calendar.

### Changes
- **Removed** hardcoded event HTML from `content/en/_index.html`
- **Added** `kubecon-events.html` shortcode for dynamic event rendering
- **Added** `data/kubecon_events.yaml` containing structured event data generated from scripts/fetch_kubecon_events
- **Added** `scripts/fetch_kubecon_events.py` to scrape and generate event data

### To Update Events
```bash
python3 scripts/fetch_kubecon_events.py
```
[Preview](https://deploy-preview-53519--kubernetes-io-main-staging.netlify.app/) 

Related to #52376

Caesarsage committed:
automate kubecon upcoming events

lmktfy reviewed (commented):
This is good, but there's a challenge: how does this interact with localization?

* eg: make sure to mark the text as left to right in case the parent page isn't
* eg: we might want to use a localized date format
* other ideas can fit too
/area web-development

lmktfy commented on scripts/fetch_kubecon_events.py:
```diff
@@ -0,0 +1,183 @@
+    """Scrape KubeCon events from Linux Foundation calendar"""
+
+    headers = {
+        'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36'
```
nit: this is mildly deceptive

singh1203 reviewed (approved):
LGTM!

@Caesarsage, Thank you. This is a solid implementation that eliminates the need for manual event updates. The filtering logic and error handling are particularly well done.

nate-double-u reviewed (commented):
Thanks for this @Caesarsage 

A concern: this may circumvent the steering committee's ability to approve site-wide changes. Usually any update to the banner needs to be approved. Tho, I suppose if this is run and generates a PR that could be approved that would alleviate the concern. 

Also, each banner has a UTM link associated with it generated by the LF Marketing team that wouldn't be captured by scraping the event page.

How would this manage scheduling the banners (and ensure there is no overlap between existing banners, or if there is, how would this decide which has precedence)?

Caesarsage commented:
> Thanks for this @Caesarsage
> 
> A concern: this may circumvent the steering committee's ability to approve site-wide changes. Usually any update to the banner needs to be approved. Tho, I suppose if this is run and generates a PR that could be approved that would alleviate the concern.
> 
> Also, each banner has a UTM link associated with it generated by the LF Marketing team that wouldn't be captured by scraping the event page.
> 
> How would this manage scheduling the banners (and ensure there is no overlap between existing banners, or if there is, how would this decide which has precedence)?

The proposed change here is a utility function that can be used to generate the data used by the shortcode  for the event list section on the https://k8.io/ . So yes someone will need to generate the data with the function and raise the PR for the changes.

And as Tim mentioned it doesn't tough the site wide banner .

dkarczmarski commented:
> If so, locations like Amsterdam or Mumbai would not be localized and would always be displayed as their original English names. I just wanted to confirm whether this matches your intention. (localized pages may end up mixing the local language with English for the location names. this could potentially hurt the user experience on localized sites)

I’m not sure if my question was understood correctly, but I was referring to exactly the same thing :)

Caesarsage committed:
Address review feedback: add i18n support, fix User-Agent, add deps

Caesarsage commented:
Hi everyone, I've addressed the feedback from the review. Here's a summary of the changes:

1. i18n support for region names (feedback from @lmktfy, @seokho-son , @dkarczmarski)

Added kubecon_region_* keys to i18n/en/en.toml for all regions (North America, Europe, India, Japan, China) plus a fallback
Updated the kubecon-events.html shortcode to translate region names using Hugo's T function via a dict mapping
City names and dates remain in English as discussed, localization teams can add translations to their respective i18n/{lang}/{lang}.toml files and adopt the {{< kubecon-events >}} shortcode in follow-up PRs
2. Honest User-Agent (feedback from @lmktfy)

Replaced the fake Chrome User-Agent with kubernetes-website-bot/1.0 (+https://github.com/kubernetes/website)
3. Rebase

4. Future consideration: CI/CD automation

If this approach is approved, I think we should add a GitHub Actions workflow (similar to the existing update-schedule.yml) that:

Runs scripts/fetch_kubecon_events.py on a schedule (e.g., weekly via cron)
Checks if data/kubecon_events.yaml has changed
Automatically opens a PR with the updated event data if changes are detected
This would complete the automation loop — the script generates the data, CI keeps it current, and the shortcode renders it. 

The PR-based approach preserves the steering committee's ability to review and approve changes before they go live, which addresses @nate-double-u's earlier concern.

lmktfy commented on data/kubecon_events.yaml:
```diff
@@ -0,0 +1,21 @@
+events:
+- name: KubeCon + CloudNativeCon Europe
+  region: Europe
+  date: Mar 23–26, 2026
```
Better: use ISO 8601 representation of dates.

seokho-son reviewed (commented):
Hi @Caesarsage 
Thanks for this improvement.
One concern I still have is around localization. While region names are now partially handled via i18n, city names and date formats are still effectively English-oriented, which may lead to mixed-language output on localized pages.

To help localization teams handle this more clearly, I suggest adding a short comment explaining:

- how {{< kubecon-events >}} works (data-driven via kubecon_events.yaml)
- current limitations around localization (city/date)
- and the option to replace it with a static, fully localized version if needed

Caesarsage committed:
format date with hugo and added localization comment for city names

Caesarsage committed:
added pl example

lmktfy reviewed (commented):
I think this is a good thing to do.

Caesarsage committed:
Revert Polish example, move data to data/events/kubecon.yaml

Caesarsage committed:
Use JSON-LD for KubeCon event data extraction and added color scheme

lmktfy reviewed (commented):
Tip: if using AI, make sure to check that the LLM output doesn't include explanations of the diff between the last and newest changes

dkarczmarski commented on i18n/pl/pl.toml:
```diff
@@ -287,7 +269,6 @@ other = "Wyszukaj polecenia lub opis ..."
 
 [kubectl_ref_view_full]
 other = "Zobacz pełną dokumentację"
-
```
the change at line 290 seems unnecessary :)

nate-double-u replied:
Agreed, we should probably not change localizations in this PR.

nate-double-u commented on scripts/fetch_kubecon_events.py:
```diff
@@ -0,0 +1,270 @@
+
+# Configuration
+CALENDAR_URL = "https://events.linuxfoundation.org/about/calendar/?_sf_s=kubecon"
+EVENT_LIMIT = 2
```
nit: why 2? If there are 3 events coming up this would skip one.

lmktfy replied:
I guess if we really want to show 3 upcoming events, change this locally and run the script with that setting.

nate-double-u reviewed (commented):
We may be able to build an API to call to get the same data in a cleaner way. Could you add the spec you want as an issue to the Events site here: https://github.com/linuxfoundation/lfevents/issues/new

Otherwise, in the meantime, this looks good to me, thanks @Caesarsage.

/lgtm

Caesarsage committed:
Restore pl.toml blank line and regenerate kubecon events

lmktfy commented on content/en/_index.html:
```diff
@@ -52,13 +52,8 @@ <h2>The Absolute Beginner's Guide To Cloud Native</h2>
-  <div>
-    <a href="https://events.linuxfoundation.org/kubecon-cloudnativecon-north-america-2026/" class="desktopKCButton"><strong>North America</strong> (Salt Lake City, Nov 9-12, 2026)</a>
-  </div>
+{{< kubecon-events >}}
```
I consider this resolved.

lmktfy commented on layouts/shortcodes/kubecon-events.html:
```diff
@@ -0,0 +1,18 @@
+{{ range $data.events -}}
+{{ $startDate := time.AsTime .start_date -}}
+{{ $endDate := time.AsTime .end_date -}}
+{{ $regionKeys := dict "North America" "kubecon_region_north_america" "Europe" "kubecon_region_europe" "India" "kubecon_region_india" "Japan" "kubecon_region_japan" "China" "kubecon_region_china" "Event" "kubecon_region_event" }}
```
nit: could put this mapping in (eg) `data/events/region_name_heuristics.yaml`

lmktfy commented on scripts/fetch_kubecon_events.py:
```diff
@@ -0,0 +1,270 @@
+    'User-Agent': 'kubernetes-website-bot/1.0 (+https://github.com/kubernetes/website)'
+}
+
+REGIONS = {
+    "North America": "North America",
+    "Europe": "Europe",
+    "India": "India",
+    "Japan": "Japan",
+    "China": "China",
+}
```
nit: this could be tidier.

lmktfy commented on scripts/fetch_kubecon_events.py:
```diff
@@ -0,0 +1,270 @@
+    """Check if title matches the expected KubeCon event pattern."""
+    if KUBECON_TITLE_MATCH_RE.match(title):
+        return True
+    print(f"  Skipped (not a main KubeCon event): {title}")
```
(nit)
```suggestion
    print(f"  Skipped (not a main KubeCon event): {title}", file=sys.stderr)
```

lmktfy commented on scripts/fetch_kubecon_events.py:
```diff
@@ -0,0 +1,270 @@
+    soup = BeautifulSoup(response.text, 'html.parser')
+    ld_event = _extract_json_ld_event(soup)
+    if not ld_event:
+        print(f"  Warning: No JSON-LD Event found at {url}")
```
(nit)
```suggestion
        print(f"  Warning: No JSON-LD Event found at {url}", file=sys.stderr)
```

lmktfy commented on scripts/fetch_kubecon_events.py:
```diff
@@ -0,0 +1,270 @@
+        events = fetch_kubecon_events()
+
+        if not events:
+            print("\nNo matching KubeCon events found!")
+            print("  Looking for: 'KubeCon + CloudNativeCon <region>'")
```
nit: could write this to stderr

lmktfy commented on scripts/fetch_kubecon_events.py:
```diff
@@ -0,0 +1,270 @@
+                print(f"     Colors: {', '.join(event['colors'])}")
+
+    except requests.exceptions.RequestException as e:
+        print(f"\nNetwork Error: {e}")
+        print("  Could not fetch the events page.")
```
nit: could write this to stderr

lmktfy commented on scripts/fetch_kubecon_events.py:
```diff
@@ -0,0 +1,270 @@
+        print("  Could not fetch the events page.")
+        exit(1)
+    except Exception as e:
+        print(f"\nError: {e}")
```
```suggestion
        print(f"\nError: {e}", file=sys.stderr)
```

lmktfy reviewed (commented):
/lgtm

**I don't see harm from merging this as-is.**

Caesarsage commented:
Could use a follow PR, especially after the https://github.com/linuxfoundation/lfevents/issues/1107

Caesarsage commented:
Cc @dipesh-rawat for approval

cjyabraham commented:
> Could use a follow PR, especially after the [linuxfoundation/lfevents#1107](https://github.com/linuxfoundation/lfevents/issues/1107)

[This PR](https://github.com/linuxfoundation/lfevents/pull/1108) is ready to address that issue.

Caesarsage committed:
Retrigger Netlify preview

reylejano commented:
In a follow-up PR, please add this script to the scripts README.md https://github.com/kubernetes/website/blob/main/scripts/README.md
/approve
